pkg/etcd/client.go

```diff
@@ -165,7 +176,53 @@ func (c *Client) TimeToLive(ctx context.Context, lease clientv3.LeaseID, opts ..
 
 // Watch delegates request to clientv3.Watcher.Watch
 func (c *Client) Watch(ctx context.Context, key string, opts ...clientv3.OpOption) clientv3.WatchChan {
-	return c.cli.Watch(ctx, key, opts...)
+	watchCh := make(chan clientv3.WatchResponse, etcdWatchChBufferSize)
+	go c.WatchWithChan(ctx, watchCh, key, opts...)
+	return watchCh
+}
+
+// WatchWithChan maintains a watchCh and send all msg from the watchCh to outCh
+func (c *Client) WatchWithChan(ctx context.Context, outCh chan clientv3.WatchResponse, key string, opts ...clientv3.OpOption) {
+	var lastRevision int64
+	ctx1, cancel := context.WithCancel(ctx)
+	defer cancel()
+	watchCh := c.cli.Watch(ctx1, key, opts...)
+	watchTimer := c.clock.Timer(etcdWatchChTimeoutDuration)
+	defer watchTimer.Stop()
+	requestTimer := c.clock.Timer(etcdRequestProgressDuration)
+	defer requestTimer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			cancel()
+			return
+		case response := <-watchCh:
+			watchTimer.Reset(etcdWatchChTimeoutDuration)
+			requestTimer.Reset(etcdRequestProgressDuration)
+			if response.Err() == nil {
+				lastRevision = response.Header.Revision
+			}
+			select {
+			case <-ctx.Done():
+				cancel()
+				return
+			case outCh <- response: // it may blocking here
+			}
+		case <-requestTimer.C:
+			if err := c.RequestProgress(ctx); err != nil {
+				log.Warn("failed to request progress for etcd watcher", zap.Error(err))
+			}
+			requestTimer.Reset(etcdRequestProgressDuration)
+		case <-watchTimer.C:
+			log.Warn("etcd watchCh blocking too long, reset the watchCh")
```

Comment:
Can you attach a stack trace in the log, so that we know how this function gets called?